### PR TITLE
feat(vod): add new resource media asset

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

add new resource media asset

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
add new resource media asset
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/vod' TESTARGS='-run TestAccMediaAsset_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vod -v -run TestAccMediaAsset_basic -timeout 360m -parallel 4
=== RUN   TestAccMediaAsset_basic
=== PAUSE TestAccMediaAsset_basic
=== CONT  TestAccMediaAsset_basic
--- PASS: TestAccMediaAsset_basic (30.37s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vod       30.479s
```

### CODE REVIEW

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/vod' TESTARGS='-run TestAccMediaAsset_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vod -v -run TestAccMediaAsset_basic -timeout 360m -parallel 4
=== RUN   TestAccMediaAsset_basic
=== PAUSE TestAccMediaAsset_basic
=== CONT  TestAccMediaAsset_basic
--- PASS: TestAccMediaAsset_basic (33.23s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vod       33.326
```

File: docs/resources/vod_media_asset.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+## Import
+
+The media asset can be imported using the `id`, e.g.
+
+```
+$ terraform import huaweicloud_vod_media_asset.test 8754976729b8a2ba745d01036edded2b
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attrubutes missing from the
+API response, security or some other reason. The missing attributes include: `url`, `input_bucket`, `input_bucket`,
```
Comment (hmbig2) [inline docs/resources/vod_media_asset.md]: Duplicate `input_bucket`

File: docs/resources/vod_media_asset.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+It is generally recommended running `terraform plan` after importing a media asset.
+You can then decide if changes should be applied to the media asset, or the resource
+definition should be updated to align with the media asset. Also you can ignore changes as below.
+
+```
+resource "huaweicloud_vod_media_asset" "test" {
+    ...
+  lifecycle {
+    ignore_changes = [
+      url, input_bucket, input_bucket, input_path, output_bucket, output_path, storage_mode, template_group_name,
```
Comment (hmbig2) [inline docs/resources/vod_media_asset.md]: ditto

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+func resourceMediaAssetUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.HcVodV1Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating VOD client: %s", err)
+	}
+
+	updateOpts := vod.UpdateAssetMetaReq{
```
Comment (hmbig2) [inline huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go]: add haschanges

File: docs/resources/vod_media_asset.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+## Import
+
+The media asset can be imported using the `id`, e.g.
+
+```
+$ terraform import huaweicloud_vod_media_asset.test 8754976729b8a2ba745d01036edded2b
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attrubutes missing from the
```
Comment (niuzhenguo) [inline docs/resources/vod_media_asset.md]: s/attrubutes/attributes

File: docs/resources/vod_media_asset.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  provider-level region will be used. Changing this creates a new resource.
+
+* `name` - (Required, String) Specifies the media asset name, which contains a maximum of 128 characters.
+
+* `media_type` - (Required, String, ForceNew) Specifies the media type. Valid values are: **MP4**, **TS**, **MOV**,
+  **MXF**, **MPG**, **FLV**, **WMV**, **AVI**, **M4V**, **F4V**, **MPEG**, **3GP**, **ASF**, **MKV**, **HLS**,
+  **M3U8**, **MP3**, **OGG**, **WAV**, **WMA**, **APE**, **FLAC**, **AAC**, **AC3**, **MMF**, **AMR**, **M4A**,
+  **M4R**, **WV**, **MP2**. Changing this creates a new resource.
+
+  -> When `video_type` is set to **HLS**, `storage_mode` must be set to **1** (user bucket), and the `output_path`
```
Comment (niuzhenguo) [inline docs/resources/vod_media_asset.md]: s/video_type/media_type

File: huaweicloud/services/acceptance/vod/resource_huaweicloud_vod_media_asset_test.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		getResourceAsset,
+	)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheckVODMediaAsset(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		CheckDestroy:      rc.CheckResourceDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMediaAsset_basic(testAccMediaAsset_base(rName), rName, "test video"),
```
Comment (niuzhenguo) [inline huaweicloud/services/acceptance/vod/resource_huaweicloud_vod_media_asset_test.go]: Nits: better to make "test video" a variable, or just hardcoded in the tf script.

File: docs/resources/vod_media_asset.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/vod_media_asset.md]: fixed

File: docs/resources/vod_media_asset.md (hunk 2 above)
Comment (Jason-Zhang9309) [reply docs/resources/vod_media_asset.md]: fixed

File: docs/resources/vod_media_asset.md (hunk 4 above)
Comment (Jason-Zhang9309) [reply docs/resources/vod_media_asset.md]: fixed

File: docs/resources/vod_media_asset.md (hunk 5 above)
Comment (Jason-Zhang9309) [reply docs/resources/vod_media_asset.md]: fixed

File: huaweicloud/services/acceptance/vod/resource_huaweicloud_vod_media_asset_test.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/vod/resource_huaweicloud_vod_media_asset_test.go]: fixed

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go (hunk 3 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go]: fixed

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/vod' TESTARGS='-run TestAccMediaAsset_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/vod -v -run TestAccMediaAsset_basic -timeout 360m -parallel 4
=== RUN   TestAccMediaAsset_basic
=== PAUSE TestAccMediaAsset_basic
=== CONT  TestAccMediaAsset_basic
--- PASS: TestAccMediaAsset_basic (40.30s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/vod       40.404s
```

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/oms' TESTARGS='-run TestAccMigrationTask_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/oms -v -run TestAccMigrationTask_basic -timeout 360m -parallel 4
=== RUN   TestAccMigrationTask_basic
=== PAUSE TestAccMigrationTask_basic
=== CONT  TestAccMigrationTask_basic
--- PASS: TestAccMigrationTask_basic (90.04s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/oms       90.140s

make testacc TEST='./huaweicloud/services/acceptance/oms' TESTARGS='-run TestAccMigrationTask_prefix'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/oms -v -run TestAccMigrationTask_prefix -timeout 360m -parallel 4
=== RUN   TestAccMigrationTask_prefix
=== PAUSE TestAccMigrationTask_prefix
=== CONT  TestAccMigrationTask_prefix
--- PASS: TestAccMigrationTask_prefix (55.86s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/oms       55.968s

make testacc TEST='./huaweicloud/services/acceptance/oms' TESTARGS='-run TestAccMigrationTask_list'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/oms -v -run TestAccMigrationTask_list -timeout 360m -parallel 4
=== RUN   TestAccMigrationTask_list
=== PAUSE TestAccMigrationTask_list
=== CONT  TestAccMigrationTask_list
--- PASS: TestAccMigrationTask_list (48.40s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/oms       48.502s
```

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+	createReq := vod.PublishAssetFromObsRequest{
+		Body: &createOpts,
+	}
+
+	resp, err := client.PublishAssetFromObs(&createReq)
+	if err != nil {
+		return "", fmt.Errorf("error creating VOD media asset: %s", err)
+	}
+
+	return *resp.AssetId, nil
```
Comment (hmbig2) [inline huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go:355]: check the point

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+	}
+	resp, err := client.UploadMetaDataByUrl(&createReq)
+	if err != nil {
+		return "", fmt.Errorf("error creating VOD media asset: %s", err)
+	}
+	if resp.UploadAssets == nil {
+		return "", fmt.Errorf("unable to find the asset after uploading")
+	}
+	assets := *resp.UploadAssets
+	return *assets[0].AssetId, nil
```
Comment (hmbig2) [inline huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go:288]: check the length and point

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go:355]: fixed

File: huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go (hunk 8 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/vod/resource_huaweicloud_vod_media_asset.go:288]: fixed

File: docs/resources/vod_media_asset.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+### Upload media asset by URL
+
+```hcl
+variable "media_url" {}
+
+resource "huaweicloud_vod_media_asset" "test" {
+  name         = "test"
+  media_type   = "MP4"
+  url          = var.media_url
+  description  = "test video"
```
Comment (ShiChangkuo) [inline docs/resources/vod_media_asset.md]: two spaces between parameters and "="

Comment (Jason-Zhang9309) [reply docs/resources/vod_media_asset.md]: fixed

**Pull request merged**